spinalcordtoolbox/scripts/sct_dmri_display_bvecs.py

```diff
@@ -52,37 +68,93 @@ def get_parser():
+    # plt.axis('equal')
     plt.xlabel(xlabel)
     plt.ylabel(ylabel)
-    plt.xlim([-1, 1])
-    plt.ylim([-1, 1])
+    plt.xlim([-max(bvals), max(bvals)])
+    plt.ylim([-max(bvals), max(bvals)])
     plt.grid()
 
+
+def create_custom_legend(fig, shell_colors, bvals):
+    """
+    Create single custom legend for whole figure
+    :param fig: figure the legend will be created for
+    :param shell_colors: dict with b-values and colors
+    :param bvals: ndarray with all b-values
+    :return:
+    """
+
+    # count number of bvals for individual shells
+    bvals_per_shell = pd.Series(bvals[bvals > bzero]).value_counts()
+
+    # Create single custom legend for whole figure with several subplots
+    lines = list()  # initialize list for individual symbols in the legend
+    labels = list()  # initialize list for individual text labels in the legend
+    # Loop across legend elements
+    for key, value in shell_colors.items():
+        lines.append(Line2D([0], [0], color=value, marker='o', markersize=8, alpha=0.7, linestyle=''))
+        labels.append('b-values = {} (n = {})'.format(str(int(key)), str(bvals_per_shell[key])))
+
+    plt.tight_layout()
+    # Insert legend below subplots, NB - this line has to be below the plt.tight_layout()
+    legend = fig.legend(lines, labels, loc='lower left', bbox_to_anchor=(0.1, 0),
+                        bbox_transform=plt.gcf().transFigure, ncol=len(lines), fontsize=10)
+    # Change box's frame color to black
+    frame = legend.get_frame()
+    frame.set_edgecolor('black')
+    # tight layout of whole figure and shift master title up
+    fig.tight_layout()
+    fig.subplots_adjust(top=0.88, bottom=0.1)
+
+
 # MAIN
 # ==========================================================================================
 
-
 def main(argv=None):
     parser = get_parser()
     arguments = parser.parse_args(argv)
     verbose = arguments.v
     set_global_loglevel(verbose=verbose)
 
     fname_bvecs = arguments.bvec
+    if arguments.bval is not None:
+        fname_bvals = arguments.bval
+        # Read bvals and bvecs files
+        bvals, bvecs = read_bvals_bvecs(fname_bvals, fname_bvecs)
+    else:
+        # Read only bvecs file
+        _, bvecs = read_bvals_bvecs(None, fname_bvecs)
 
-    # Read bvecs
-    bvecs = read_bvals_bvecs(fname_bvecs, None)
-    bvecs = bvecs[0]
     # if first dimension is not equal to 3 (x,y,z), transpose bvecs file
     if not bvecs.shape[0] == 3:
         bvecs = bvecs.transpose()
-    x, y, z = bvecs[0], bvecs[1], bvecs[2]
+
+    # bvals file was passed
+    if arguments.bval is not None:
+        # multiply unit b-vectors by b-values
+        x, y, z = bvecs[0] * bvals, bvecs[1] * bvals, bvecs[2] * bvals
+    # bvals file was not passed
+    else:
+        x, y, z = bvecs[0], bvecs[1], bvecs[2]
+        # create dummy unit bvals array (necessary fot scatter plots)
+        bvals = np.repeat(1, bvecs.shape[1])
```

Comment:
I am not sure, because there has to be check for bvecs transposition:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/a6886da7caa1efb329701d0493bde9c0005162fd/spinalcordtoolbox/scripts/sct_dmri_display_bvecs.py#L130-L131

Reply:
Ah, good point. Hrm.... Well, could that statement be moved too?

I also realize that this check could be simplified as well:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/a6886da7caa1efb329701d0493bde9c0005162fd/spinalcordtoolbox/scripts/sct_dmri_display_bvecs.py#L133-L141

Since for the `else:` case, if `bvals` becomes a dummy array of 1s, then multiplying would be the same as `bvecs[0], bvecs[1], bvecs[2]`, I think? So perhaps all together it could look something like:

```python
    fname_bvecs = arguments.bvec
    fname_bvals = arguments.bval
    bvals, bvecs = read_bvals_bvecs(fname_bvals, fname_bvecs)
    # if first dimension is not equal to 3 (x,y,z), transpose bvecs file
    if bvecs.shape[0] != 3: 
        bvecs = bvecs.transpose() 
    # create dummy unit bvals array if bvals wasn't specified (necessary for scatter plots)
    if bvals is None:
        bvals = np.repeat(1, bvecs.shape[1])
    # multiply unit b-vectors by b-values
    x, y, z = bvecs[0] * bvals, bvecs[1] * bvals, bvecs[2] * bvals

Reply:
> Since for the else: case, if bvals becomes a dummy array of 1s, then multiplying would be the same as bvecs[0], bvecs[1], bvecs[2], I think?

@joshuacwnewton of course, you are right! :-) Refactored in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3390/commits/057227447e8924fc214bfc211437cba0870e70c9